File: Programming - 2/Lab_5_6/Lab_6_CPP/Back/Triangle.cpp
#include <cmath>
#include "Triangle.h"

Triangle::Triangle(Point2D A, Point2D B, Point2D C)
{
    if ((A.x == B.x && B.x == C.x) || (A.y == B.y && B.y == C.y))
    {
        throw "Vertexes can't be on the same line";
    }
    _vertexA = A;
    _vertexB = B;
    _vertexC = C;
}

Point2D Triangle::GetVertexA() { return _vertexA; }
Point2D Triangle::GetVertexB() { return _vertexB; }
Point2D Triangle::GetVertexC() { return _vertexC; }

double Triangle::GetSideA() { return GetDistance(_vertexB, _vertexC); }
double Triangle::GetSideB() { return GetDistance(_vertexA, _vertexC); }
double Triangle::GetSideC() { return GetDistance(_vertexA, _vertexB); }

double Triangle::GetPerimeter() { return GetSideA() + GetSideB() + GetSideC(); }
double Triangle::GetArea()
{
    return  sqrt((GetPerimeter() / 2) * (GetPerimeter() / 2 - GetSideA()) * 
    (GetPerimeter() / 2 - GetSideB()) * (GetPerimeter() / 2 - GetSideC()));
}

File: Programming - 2/Lab_5_6/Shapes/Triangle.cs
using System;

namespace Shapes
{
    public sealed class Triangle : Shape
    {
        public Triangle((double x, double y) A, (double x, double y) B, (double x, double y) C)
        {
            if ((A.x == B.x && B.x == C.x) || (A.y == B.y && B.y == C.y))
            {
                throw new ArgumentException("Vertexes can't be on the same line");
            }
            VertexA = A;
            VertexB = B;
            VertexC = C;
        }

        public (double x, double y) VertexA { get; private set; }
        public (double x, double y) VertexB { get; private set; }
        public (double x, double y) VertexC { get; private set; }

        public double SideA => GetDistance(VertexB, VertexC);
        public double SideB => GetDistance(VertexA, VertexC);
        public double SideC => GetDistance(VertexA, VertexB);

        public override double Area 
            => Math.Sqrt((Perimeter / 2) * (Perimeter / 2 - SideA) * (Perimeter / 2 - SideB) * (Perimeter / 2 - SideC));
        public override double Perimeter => SideA + SideB + SideC;
    }
}
